programs/editor/Tools.js

```diff
@@ -62,24 +62,32 @@ define("webodf/editor/Tools", [
                 aboutDialog,
                 sessionSubscribers = [];
 
+            function placeAndStartUpWidget(widget) {
+                widget.placeAt(toolbar);
+                widget.startup();
+            }
+
             /**
              * Creates a tool and installs it, if the enabled flag is set to true.
              * Only supports tool classes whose constructor has a single argument which
              * is a callback to pass the created widget object to.
              * @param {!function(new:Object, function(!Object):undefined)} Tool  constructor method of the tool
              * @param {!boolean} enabled
+             * @param {!Object|undefined=} config
              * @return {?Object}
              */
-            function createTool(Tool, enabled) {
+            function createTool(Tool, enabled, config) {
                 var tool = null;
 
                 if (enabled) {
-                    tool = new Tool(function (widget) {
-                        widget.placeAt(toolbar);
-                        widget.startup();
-                    });
+                    if (config) {
+                        tool = new Tool(config, placeAndStartUpWidget);
```

Comment:
Using a callback in constructor is bad if the callback if object initialization is asynchronous.
An example of a bad callback is OdfContainer. Initialization is async there and can fail. If it fails, no OdfContainer object should have been returned. In such cases a factory function with a callback is preferred.
Also in this case it seems that the tool is not useable until placeAndStartUpWidget returns.